jax/_src/custom_derivatives.py

```diff
@@ -553,8 +554,10 @@ def _flatten_bwd(in_tree, in_avals, out_trees, *args):
            "number of arguments to the primal function, but got VJP output "
            "structure {} for primal input structure {}.")
     raise TypeError(msg.format(in_tree2, in_tree)) from None
-  yield [zeros_like_aval(aval.at_least_vspace()) if ct is zero else ct
-         for aval, ct in zip(in_avals, cts_in_flat)]
+  # Ignore any None cotangents, and any corresponding to inputs for which the
+  # type doesn't equal the tangent type (i.e. float0s)
+  yield [Zero(a.at_least_vspace()) if ct is zero or a != a.at_least_vspace()
```

Comment:
The `or a != a.at_least_vspace()` is the only thing we needed to fix the bug at hand.

Another change on this line was to produce symbolic zeros using `Zero` rather than `zeros_like_aval`. I couldn't resist making this optimization, even though it is unrelated. (I hope it doesn't cause unexpected breakages...)